internal/provider/resource_policy_attachment.go

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package provider
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/hashicorp/boundary/api"
+	"github.com/hashicorp/boundary/api/scopes"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const (
+	policyPolicyIdKey         = "policy_id"
+	policyStoragePolicyPrefix = "pst_"
+)
+
+func resourcePolicyAttachment() *schema.Resource {
+	return &schema.Resource{
+		CreateContext:        resourcePolicyAttachmentCreate,
+		ReadWithoutTimeout:   resourcePolicyAttachmentRead,
+		DeleteWithoutTimeout: resourcePolicyAttachmentDelete,
+
+		Importer: &schema.ResourceImporter{
+			StateContext: schema.ImportStatePassthroughContext,
+		},
+
+		Schema: map[string]*schema.Schema{
+			ScopeIdKey: {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			policyPolicyIdKey: {
+				Type:     schema.TypeString,
+				ForceNew: true,
+				Required: true,
+			},
+		},
+	}
+}
+
+func resourcePolicyAttachmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	md := meta.(*metaData)
+	scp := scopes.NewClient(md.client)
+	opts := []scopes.Option{scopes.WithAutomaticVersioning(true)}
+
+	var scopeId string
+	if scopeIdVal, ok := d.GetOk(ScopeIdKey); ok {
+		scopeId = scopeIdVal.(string)
+	} else {
+		return diag.Errorf("no scope ID provided")
+	}
+
+	if err := d.Set(ScopeIdKey, scopeId); err != nil {
+		return diag.FromErr(err)
+	}
+
+	var policyId string
+	if policyIdVal, ok := d.GetOk(policyPolicyIdKey); ok {
+		policyId = policyIdVal.(string)
+	} else {
+		return diag.Errorf("no policy ID provided")
+	}
+
+	var policy_type string
+	switch {
+	case strings.HasPrefix(policyId, policyStoragePolicyPrefix):
+		if _, err := scp.AttachStoragePolicy(ctx, scopeId, 0, policyId, opts...); err != nil {
+			return diag.FromErr(err)
+		}
+		policy_type = "storage"
+	default:
+		return diag.Errorf("unknown policy type provided.")
+	}
+
+	if err := d.Set(policyPolicyIdKey, policyId); err != nil {
+		// TODO: does this ordering prevent tf from storing the tfstate if the attach fails?
+		return diag.FromErr(err)
+	}
+
+	// scopes can have one and only one policy of each type
```

Comment:
This should be able to be updated to scopeId with the subtype specific resource change

Reply:
I think that the policy attachment should stay subtype-agnostic

Reply:
I don't think that there's a reason to need to add new policy type attachments for each new type. unlike the actual policy definition, there's no risk of schema bloat or anything with the attachment specifically.

Reply:
I do not believe we can make the assertion that future policies will be limited to one and only one. The plan is to allow multiple policies possibly of the same type, for storage we know we only want to limit it to one and only one (at least for now)

Reply:
So there are three things that make me question if this can be subtype-agnostic:
- Having `boundary_policy_storage_attachment` in the Terraform file means it becomes human readable on what the attachment is, otherwise someone needs to know that the prefix indicates it is a storage type. 
- Future policies could allow multiple of that policy type to be associated with a scope
- We also cannot guarantee that other policies wont require additional fields when associating them with a scope, for example with Storage policies we have override behavior as part of the policy - but that might not always be the case.

I am not sure I see much of a benefit in locking into a subtype agnostic approach.

Reply:
Might be good to get @AdamBouhmad thoughts on this point. Which of the following do you prefer, noting some of my concerns above:

```
resource "boundary_scope_policy_attachment" "foo" {
  scope_id  = boundary_scope.org.id
  policy_id = boundary_storage_policy.boo.id
}
```
or
```
resource "boundary_scope_policy_storage_attachment" "foo" {
  scope_id  = boundary_scope.org.id
  policy_id = boundary_storage_policy.boo.id
}
```

We can always migrate this once we have more information about future policies and how those would be assocaited with other resources.

Reply:
while waiting for adam's thoughts, my response was going to be that having the policy variable be `policy_id = boundary_policy_storage.foo.id` would make it clear enough what type of policy you're attaching. in the mean time I've changed the resource to `boundary_scope_policy_attachment`

Reply:
I think personally I prefer the former approach, given that storage policies are a subset of governance policies, and a scope in the future may have multiple policies associated((e.g a future policy could enforce session recording at the org level, and have that be enforced on all SSH targets in child scopes). There's a very similar pattern with [terraform iam policies](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_user_policy_attachment), where you'd define your iam policies, and attach them to a particular user

example:

`resource "aws_iam_user_policy_attachment" "s3" {
  user       = aws_iam_user.example_user.name
  policy_arn = aws_iam_policy.policy_s3.arn
}

resource "aws_iam_user_policy_attachment" "dynamodb" {
  user       = aws_iam_user.example_user.name
  policy_arn = aws_iam_policy.policy_dynamodb.arn
}`